src/sentry/nodestore/filesystem/backend.py

```diff
@@ -1,53 +1,58 @@
-import os
-from datetime import datetime, timedelta, timezone
+import datetime
+from io import BytesIO
 
 from django.conf import settings
 
+from sentry import options as options_store
+from sentry.models.file import get_storage
 from sentry.nodestore.base import NodeStorage
 
 
 class FileSystemNodeStorage(NodeStorage):
     """
-    A simple backend that saves each node as a file. Only appropriate for
-    debugging and development!
+    A file system-based backend that saves each node as a file, Local fileSystem only appropriate for
+    debugging and development! For production environments, please use S3 or GCS. config.yml Configuration reference:
+    >>> filestore.backend: "s3"
+    ... filestore.options:
+    ...   access_key: "xxx"
+    ...   secret_key: "xxx"
+    ...   endpoint_url: "https://s3.us-east-1.amazonaws.com"
+    ...   bucket_name: "sentry"
+    ...   location: "/sentry"
     """
 
-    def __init__(self, path: str | None = None):
-        self.path: str = ""
-
-        if not settings.DEBUG:
-            raise ValueError("FileSystemNodeStorage should only be used in development!")
-        if path:
-            self.path = os.path.abspath(os.path.expanduser(path))
-        else:
-            self.path = os.path.abspath(os.path.join(os.path.dirname(__file__), "./nodes"))
-
-    def _get_bytes(self, id: str) -> bytes:
-        with open(self.node_path(id), "rb") as file:
-            return file.read()
-
-    def _set_bytes(self, id: str, data: bytes, ttl: timedelta | None = None) -> None:
-        with open(self.node_path(id), "wb") as file:
-            file.write(data)
-
-    def delete(self, id: str) -> None:
-        os.remove(self.node_path(id))
-
-    def cleanup(self, cutoff: datetime) -> None:
-        for filename in os.listdir(self.path):
-            path = os.path.join(self.path, filename)
-            creation_datetime = datetime.fromtimestamp(os.path.getctime(path)).replace(
-                tzinfo=timezone.utc
-            )
-
-            if creation_datetime > cutoff:
-                os.remove(path)
-
-    def bootstrap(self) -> None:
-        try:
-            os.mkdir(self.path)
-        except FileExistsError:
-            pass
-
-    def node_path(self, id: str) -> str:
-        return os.path.join(self.path, f"{id}.json")
+    def __init__(self, prefix_path=None):
+        self.prefix_path: str = "nodestore/"
+        if not settings.DEBUG and options_store.get("filestore.backend") == "filesystem":
+            raise ValueError("Local fileSystem should only be used in development!")
```

Comment:
This would break existing self-hosted users that don't use (or don't have access to) S3 compatible services. This should just log a warning instead of throwing a ValueError.

Reply:
I don't think so. The default `NodeStore` is `sentry.nodestore.django.DjangoNodeStorage` rather than this. So it won't break any existing users. If you want to configure this `FileSystemNodeStorage` in non-debug environment, you must configure `FileStore` also.

Reply:
I'd like to see this merged. 
Can we settle with being more backward compatible?

i.e. do not make the assumption that "If you want to configure this FileSystemNodeStorage in non-debug environment, you must configure FileStore also."

Reply:
Ping to bump this PR.

I would really like to see the built-in support for S3 Node Store (even though performance is not endorsed).

I think the concern here is:   `FileSystemNodeStorage` by title is supposed to be using "File System". It is not intuitive that by enabling "S3 File Store", FileSystemNodeStore will push things to S3.

I think better way is to explicitly create a new S3NodeStore class with those logic.

Reply:
See if #82125 makes sense to everyone.